src/ucp/proto/proto_reconfig.c

```diff
@@ -0,0 +1,111 @@
+     * the remote key
+     */
+    prev_rkey_cfg_index = req->send.proto_config->rkey_cfg_index;
+    if (prev_rkey_cfg_index == UCP_WORKER_CFG_INDEX_NULL) {
+        proto_select   = &worker->ep_config[ep->cfg_index].proto_select;
+        rkey_cfg_index = UCP_WORKER_CFG_INDEX_NULL;
+    } else {
+        rkey_config_key = worker->rkey_config[prev_rkey_cfg_index].key;
+        rkey_config_key.ep_cfg_index = ep->cfg_index;
+
+        status = ucp_worker_get_rkey_config(worker, &rkey_config_key,
+                                            &rkey_cfg_index);
+        if (status != UCS_OK) {
+            ucs_error("failed to switch to new rkey");
+            return UCS_OK;
+        }
+
+        proto_select = &worker->rkey_config[rkey_cfg_index].proto_select;
+    }
+
+    /* Select from protocol hash according to saved request parameters */
+    status = ucp_proto_request_set_proto(worker, ep, req, proto_select,
+                                         rkey_cfg_index,
+                                         &req->send.proto_config->select_param,
+                                         req->send.state.dt_iter.length);
+    if (status != UCS_OK) {
+        /* will try again later */
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    return req->send.uct.func(&req->send.uct);
+}
+
+static ucs_status_t ucp_proto_reconfig_progress(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_ep_h ep        = req->send.ep;
+
+    if (ep->cfg_index != req->send.proto_config->ep_cfg_index) {
+        ucp_trace_req(req,
+                      "ep configuration changed from %d to %d,"
+                      " reconfigure protocol",
+                      req->send.proto_config->ep_cfg_index, ep->cfg_index);
+        return ucp_proto_reconfig_select_progress(self);
+    }
+
+    /* TODO select wireup lane when needed */
+    req->send.lane = ucp_ep_config(ep)->key.am_lane;
+    return UCS_ERR_NO_RESOURCE;
+}
+
+static ucs_status_t
+ucp_proto_reconfig_init(const ucp_proto_init_params_t *init_params)
+{
+    /* Default reconfiguration protocol is a fallback for any case protocol
+     * selection for API functions is unsuccessful. The protocol keeps queuing
+     * requests until they can be executed.
+     * For internal operations (e.g rendezvous response) this protocol is not
+     * supported.
+     */
+    *init_params->priv_size                 = 0;
+    init_params->caps->cfg_thresh           = UCS_MEMUNITS_INF;
+    init_params->caps->cfg_priority         = 0;
+    init_params->caps->min_length           = 0;
+    init_params->caps->num_ranges           = 1;
+    init_params->caps->ranges[0].max_length = SIZE_MAX;
+    init_params->caps->ranges[0].perf       = ucs_linear_func_make(INFINITY, 0);
+
+    return UCS_OK;
+}
+
+static void ucp_proto_reconfig_str(const void *priv, ucs_string_buffer_t *strb)
+{
+    ucs_string_buffer_init(strb);
+}
+
+static ucp_proto_t ucp_reconfig_proto = {
```

Comment:
how/when it will be selected?

Reply:
added comment in https://github.com/openucx/ucx/pull/6185/files#diff-a276828aa45fa267d670e1a6817dde5a6195b2b43a443777757558a112faa9acR85